### PR TITLE
types: improve the code coverage

### PR DESCRIPTION
**Pull request by xiekeyi98**
### What problem does this PR solve? 

Improve code coverage in the package types.

### What is changed and how it works?

Add more tests to cover more functions.

### Check List 

Tests 

 - Unit test
 - Integration test

### CODE REVIEW

**Comment by xiekeyi98** [comment]
```
xiekeyi@ubuntu:~/gopath/src/github.com/pingcap/tidb/types$ go test -cover | grep "cover"
coverage: 79.3% of statements
xiekeyi@ubuntu:~/gopath/src/github.com/pingcap/tidb/types$ git checkout improveCoverage 
切换到分支 'improveCoverage'
您的分支与上游分支 'origin/improveCoverage' 一致。
xiekeyi@ubuntu:~/gopath/src/github.com/pingcap/tidb/types$ go test -cover | grep "cover"
coverage: 83.4% of statements
xiekeyi@ubuntu:~/gopath/src/github.com/pingcap/tidb/types$ 
```

File: types/etc_test.go [hunk 1]
```diff
@@ -186,10 +190,154 @@ func (s *testTypeEtcSuite) TestTruncate(c *C) {
...
+	c.Assert(res, Equals, false)
+
+	in.Collate = charset.CollationBin
+	res = IsBinaryStr(&in)
+	c.Assert(res, Equals, false)
+
+	in.Tp = mysql.TypeBlob
+	res = IsBinaryStr(&in)
+	c.Assert(res, Equals, true)
+
```
Comment (crazycs520) [inline types/etc_test.go]: remove this blank.

Comment (xiekeyi98) [reply types/etc_test.go]:
```suggestion
```

File: types/etc_test.go [hunk 2]
```diff
@@ -186,10 +190,154 @@ func (s *testTypeEtcSuite) TestTruncate(c *C) {
...
+func (s *testTypeEtcSuite) TestIsTypePrefixable(c *C) {
+	defer testleak.AfterTest(c)()
+
+	res := IsTypePrefixable('t')
+	c.Assert(res, Equals, false)
+
+	res = IsTypePrefixable(mysql.TypeBlob)
+	c.Assert(res, Equals, true)
+
```
Comment (crazycs520) [inline types/etc_test.go]: ditto.

File: types/time_test.go [hunk 3]
```diff
@@ -1070,3 +1072,230 @@ func (s *testTimeSuite) TestCheckTimestamp(c *C) {
...
+		Time: types.ZeroTime,
+		Type: mysql.TypeTimestamp,
+		Fsp:  types.DefaultFsp,
+	}
+	c.Assert(in.InvalidZero(), Equals, true)
+	in.Time = types.FromDate(2019, 00, 00, 00, 00, 00, 00)
+	c.Assert(in.InvalidZero(), Equals, true)
+	in.Time = types.FromDate(2019, 04, 12, 12, 00, 00, 00)
+	c.Assert(in.InvalidZero(), Equals, false)
+
```
Comment (crazycs520) [inline types/time_test.go]: ditto.

File: types/time_test.go [hunk 4]
```diff
@@ -1070,3 +1072,230 @@ func (s *testTimeSuite) TestCheckTimestamp(c *C) {
...
+	for _, col := range tbl {
+		res, err := types.ExtractDurationNum(&in, col.unit)
+		c.Assert(err, IsNil)
+		c.Assert(res, Equals, col.expect)
+	}
+	res, err := types.ExtractDurationNum(&in, "TEST_ERROR")
+	c.Assert(res, Equals, int64(0))
+	c.Assert(err, ErrorMatches, "invalid unit.*")
+
```
Comment (crazycs520) [inline types/time_test.go]: ditto.

File: types/etc_test.go (hunk 2 above)
Comment (xiekeyi98) [reply types/etc_test.go]:
```suggestion
```

File: types/time_test.go (hunk 3 above)
Comment (xiekeyi98) [reply types/time_test.go]:
```suggestion
```

File: types/time_test.go (hunk 4 above)
Comment (xiekeyi98) [reply types/time_test.go]:
```suggestion
```

File: types/etc_test.go [hunk 5]
```diff
@@ -186,10 +190,152 @@ func (s *testTypeEtcSuite) TestTruncate(c *C) {
...
+	res = IsTemporalWithDate(mysql.TypeDate)
+	c.Assert(res, Equals, true)
+
+	res = IsTemporalWithDate(mysql.TypeTimestamp)
+	c.Assert(res, Equals, true)
+
+	res = IsTemporalWithDate('t')
+	c.Assert(res, Equals, false)
+
```
Comment (crazycs520) [inline types/etc_test.go]: ditto.

Comment (xiekeyi98) [reply types/etc_test.go]:
```suggestion
```

**Pull request merged**